types/linkifyjs/linkifyjs-tests.tsx

```diff
@@ -0,0 +1,188 @@
+
+    linkifyHtml("github.com", {
+        className: "new-link--url"
+    });
+
+    linkifyHtml("github.com", {
+        className(href, type) {
+            return "new-link--" + type;
+        }
+    });
+
+    linkifyHtml("github.com", {
+        className: {
+            url: "new-link--url",
+            email(href: string) {
+                return "new-link--email";
+            }
+        }
+    });
+
+    /* events */
+
+    linkifyHtml("", {
+        events: {
+            click(e) {
+                alert("Link clicked!");
+            },
+            mouseover(e) {
+                alert("Link hovered!");
+            }
+        }
+    });
+
+    /* defaultProtocol */
+
+    /* format */
+
+    linkifyHtml("", {
+        format(value, type) {
+            if (type === "url" && value.length > 50) {
+                value = value.slice(0, 50) + "…";
+            }
+            return value;
+        }
+    });
+
+    linkifyHtml("", {
+        format: {
+            url(value) {
+                return value.length > 50 ? value.slice(0, 50) + "…" : value;
+            }
+        }
+    });
+
+    /* formatHref */
+
+    linkifyHtml("This site is #rad", {
+        formatHref(href, type) {
+            if (type === "hashtag") {
+                href = "https://twitter.com/hashtag/" + href.substring(1);
+            }
+            return href;
+        }
+    });
+
+    linkifyHtml("Hey @dhh, check out issue #23", {
+        formatHref: {
+            mention(href) {
+                return "https://github.com" + href;
+            },
+            ticket(href) {
+                return (
+                    "https://github.com/SoapBox/linkifyjs/issues/" +
+                    href.substring(1)
+                );
+            }
+        }
+    });
+
+    /* ignoreTags */
+
+    linkifyHtml(
+        // tslint:disable-next-line:prefer-template
+        'Please ignore <script>var a = {}; a.com = "Hi";</script> \n' +
+            "but do <span>b.ca</span>",
+        {
+            ignoreTags: ["script", "style"]
+        }
+    );
+
+    /* nl2br */
+
+    /* tagName */
+
+    linkifyHtml("github.com", {
+        tagName: "span"
+    });
+
+    linkifyHtml("#swag", {
+        tagName: {
+            hashtag: "span"
+        }
+    });
+
+    /* target */
+
+    linkifyHtml("github.com", {
+        target: "_parent"
+    });
+
+    linkifyHtml("test-email@example.com", {
+        target: {
+            url: "_parent",
+            email: null
+        }
+    });
+
+    /* validate */
+
+    // Don't linkify links that don't begin in a protocol
+    // e.g., "http://google.com" will be linkified, but "google.com" will not.
+    linkifyHtml("www.google.com", {
+        validate: {
+            url(value) {
+                return /^(http|ftp)s?:\/\//.test(value);
+            }
+        }
+    });
+});
+
+describe("linkifyjs/react", () => {
+    // From the docs here: https://soapbox.github.io/linkifyjs/docs/linkify-react.html
```

Comment:
The point of renaming the file into `.tsx` was so that you could copy the examples verbatim :p

I was referring to adding this!

```tsx
/* Usage */

var options: LinkifyOptions = {};
var content = 'For help with GitHub.com, please email support@github.com';
<Linkify tagName="p" options={options}>{content}</Linkify>;

/* Events */

let linkProps = {
  onClick: (event) => {
    if (!confirm('Are you sure you want to leave this page?')) {
       event.preventDefault()
    }
  }
};
<Linkify options={{attributes: linkProps}}>
{null}
</Linkify>;

```

Please add this in, and also move the `// From the docs here: https://soapbox.github.io/linkifyjs/docs/linkify-react.html` comment to  _just above_ these -- this just makes it easier for people to figure out where the examples are copied from!

Reply:
So after I added the example I realised the type ofr `LinkifyOptions` does not quite match the runtime behavior of the library.

The fact that the value of `onClick` in this example is `(event: any) => void` contradicts every other example I could find in their docs (and even unit-tests), so I am not sure how to type that other than use `any` here:

```
attributes?: PossiblyFuncOfHrefAndType<{
    [attrName: string]: any;
}> | null;
```

Let me know if you have a better idea.

Reply:
I could also do... (but this is pretty bad design in my opinion)

```
attributes?: PossiblyFuncOfHrefAndType<{
    [attrName: string]: string | ((event: any) => void);
}> | null;
```

Reply:
These attributes are fed directly into the React component without any validation so realistically they can be anything. I could use `unknown` but that requires TypeScript 3.1 I think and would make the declaration more difficult to use.

Reply:
> These attributes are fed directly into the React component without any validation

In that case, I think we should probably use the same type that the React component takes in:

```tsx
    attributes?: PossiblyFuncOfHrefAndType<
        React.AnchorHTMLAttributes<HTMLAnchorElement>
    > | null;
```

Tested this out and I think it works? (`linkifyjs-tests.tsx` has no errors, but try to add a nonexistent attribute and you get an error, as expected.)

Reply:
Also, just a reminder to copy in the JSX tests!

Reply:
> Also, just a reminder to copy in the JSX tests!

I did copy in the JSX tests, not sure what you mean. I can't use the JSX syntax since the project doesn't allow it, and I can't copy the examples verbatim due to linting (no var allowed, no duplicate identifiers, etc.).

Reply:
The project doesn't allow JSX? How did stuff like this get in here?

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ink/ink-tests.tsx#L15-L17

Also you can copy it very close to verbatim:
- Replace `var` with `let`
- Put each example it its own scope

```tsx

/* Usage */

{
  let options: LinkifyOptions = {};
  let content = 'For help with GitHub.com, please email support@github.com';
  <Linkify tagName="p" options={options}>{content}</Linkify>;
}

/* Events */

{
  let linkProps = {
    onClick: (event) => {
      if (!confirm('Are you sure you want to leave this page?')) {
        event.preventDefault()
      }
    }
  };
  <Linkify options={{attributes: linkProps}}>
  {null}
  </Linkify>;
}
```

Reply:
Thanks for the block idea, I don't think I've ever actually used that until today 😄